File: test/session.test.js
const { describe, it, beforeEach } = require('node:test');
const assert = require('node:assert/strict');
const session = require('../src/session');

describe('session', () => {
  const CH = 'C_TEST';

  beforeEach(() => {
    session.end(CH);
  });

  describe('start', () => {
    it('creates a session with shuffled order', () => {
      const sess = session.start(CH, ['a', 'b', 'c']);
      assert.strictEqual(sess.channelId, CH);
      assert.strictEqual(sess.current, 0);
      assert.strictEqual(sess.order.length, 3);
      assert.deepStrictEqual(sess.order.sort(), ['a', 'b', 'c']);
    });

    it('is retrievable via get', () => {
      session.start(CH, ['a']);
      const sess = session.get(CH);
      assert.ok(sess);
      assert.deepStrictEqual(sess.order, ['a']);
    });
  });

  describe('get', () => {
    it('returns null for unknown channel', () => {
      assert.strictEqual(session.get('C_UNKNOWN'), null);
    });
  });

  describe('next', () => {
    it('advances the current index', () => {
      session.start(CH, ['a', 'b', 'c']);
      const sess = session.next(CH);
      assert.ok(sess);
      assert.strictEqual(sess.current, 1);
    });

    it('returns null when standup is over', () => {
      session.start(CH, ['a']);
      const result = session.next(CH);
      assert.strictEqual(result, null);
      assert.strictEqual(session.get(CH), null);
    });

    it('returns null for unknown channel', () => {
      assert.strictEqual(session.next('C_UNKNOWN'), null);
    });
  });

  describe('add', () => {
    it('inserts user after current speaker', () => {
      session.start(CH, ['a', 'b']);
      // Force a known order for testing
      const sess = session.get(CH);
      sess.order = ['a', 'b'];
      session.add(CH, 'c');
      assert.deepStrictEqual(sess.order, ['a', 'c', 'b']);
    });

    it('is a no-op if user already in remaining order', () => {
      const sess = session.start(CH, ['a', 'b']);
      sess.order = ['a', 'b'];
      session.add(CH, 'b');
      assert.deepStrictEqual(sess.order, ['a', 'b']);
    });

    it('returns null for unknown channel', () => {
      assert.strictEqual(session.add('C_UNKNOWN', 'x'), null);
    });
  });

  describe('remove', () => {
    it('removes user from remaining order', () => {
      const sess = session.start(CH, ['a', 'b', 'c']);
      sess.order = ['a', 'b', 'c'];
      session.remove(CH, 'b');
      assert.deepStrictEqual(sess.order, ['a', 'c']);
    });

    it('returns null when removing the last remaining user', () => {
      const sess = session.start(CH, ['a']);
      sess.order = ['a'];
      const result = session.remove(CH, 'a');
      assert.strictEqual(result, null);
      assert.strictEqual(session.get(CH), null);
    });

    it('is a no-op if user not found', () => {
      const sess = session.start(CH, ['a', 'b']);
      sess.order = ['a', 'b'];
      session.remove(CH, 'z');
      assert.deepStrictEqual(sess.order, ['a', 'b']);
    });

    it('does not remove already-spoken users', () => {
      const sess = session.start(CH, ['a', 'b', 'c']);
      sess.order = ['a', 'b', 'c'];
      session.next(CH); // current = 1 (b is speaking)
      session.remove(CH, 'a'); // a already spoke — should be no-op
      assert.deepStrictEqual(sess.order, ['a', 'b', 'c']);
    });
  });

  describe('end', () => {
    it('removes the session', () => {
      session.start(CH, ['a']);
      session.end(CH);
      assert.strictEqual(session.get(CH), null);
    });

    it('is a no-op for unknown channel', () => {
      session.end('C_UNKNOWN'); // should not throw
    });
  });
});

File: internal/session/session_test.go
package session

import (
	"reflect"
	"sort"
	"testing"
)

func noShuffle(a []string) {} // identity — preserves input order

func newTestStore() *Store {
	return NewStoreWithShuffle(noShuffle)
}

const ch = "C_TEST"

func TestStart(t *testing.T) {
	t.Run("creates a session with shuffled order", func(t *testing.T) {
		s := NewStore() // real shuffle
		sess := s.Start(ch, []string{"a", "b", "c"})
		if sess.ChannelID != ch {
			t.Errorf("ChannelID = %q, want %q", sess.ChannelID, ch)
		}
		if sess.Current != 0 {
			t.Errorf("Current = %d, want 0", sess.Current)
		}
		if len(sess.Order) != 3 {
			t.Errorf("Order length = %d, want 3", len(sess.Order))
		}
		sorted := make([]string, len(sess.Order))
		copy(sorted, sess.Order)
		sort.Strings(sorted)
		if !reflect.DeepEqual(sorted, []string{"a", "b", "c"}) {
			t.Errorf("sorted Order = %v, want [a b c]", sorted)
		}
	})

	t.Run("is retrievable via Get", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a"})
		sess := s.Get(ch)
		if sess == nil {
			t.Fatal("Get returned nil")
		}
		if !reflect.DeepEqual(sess.Order, []string{"a"}) {
			t.Errorf("Order = %v, want [a]", sess.Order)
		}
	})
}

func TestGet(t *testing.T) {
	t.Run("returns nil for unknown channel", func(t *testing.T) {
		s := newTestStore()
		if s.Get("C_UNKNOWN") != nil {
			t.Error("expected nil for unknown channel")
		}
	})
}

func TestNext(t *testing.T) {
	t.Run("advances the current index", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a", "b", "c"})
		sess := s.Next(ch)
		if sess == nil {
			t.Fatal("Next returned nil")
		}
		if sess.Current != 1 {
			t.Errorf("Current = %d, want 1", sess.Current)
		}
	})

	t.Run("returns nil when standup is over", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a"})
		result := s.Next(ch)
		if result != nil {
			t.Error("expected nil when standup is over")
		}
		if s.Get(ch) != nil {
			t.Error("session should be deleted")
		}
	})

	t.Run("returns nil for unknown channel", func(t *testing.T) {
		s := newTestStore()
		if s.Next("C_UNKNOWN") != nil {
			t.Error("expected nil for unknown channel")
		}
	})
}

func TestAdd(t *testing.T) {
	t.Run("inserts user after current speaker", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a", "b"})
		s.Add(ch, "c")
		sess := s.Get(ch)
		if !reflect.DeepEqual(sess.Order, []string{"a", "c", "b"}) {
			t.Errorf("Order = %v, want [a c b]", sess.Order)
		}
	})

	t.Run("is a no-op if user already in remaining order", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a", "b"})
		s.Add(ch, "b")
		sess := s.Get(ch)
		if !reflect.DeepEqual(sess.Order, []string{"a", "b"}) {
			t.Errorf("Order = %v, want [a b]", sess.Order)
		}
	})

	t.Run("returns nil for unknown channel", func(t *testing.T) {
		s := newTestStore()
		if s.Add("C_UNKNOWN", "x") != nil {
			t.Error("expected nil for unknown channel")
		}
	})
}

func TestRemove(t *testing.T) {
	t.Run("removes user from remaining order", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a", "b", "c"})
		s.Remove(ch, "b")
		sess := s.Get(ch)
		if !reflect.DeepEqual(sess.Order, []string{"a", "c"}) {
			t.Errorf("Order = %v, want [a c]", sess.Order)
		}
	})

	t.Run("returns nil when removing the last remaining user", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a"})
		result := s.Remove(ch, "a")
		if result != nil {
			t.Error("expected nil when last user removed")
		}
		if s.Get(ch) != nil {
			t.Error("session should be deleted")
		}
	})

	t.Run("is a no-op if user not found", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a", "b"})
		s.Remove(ch, "z")
		sess := s.Get(ch)
		if !reflect.DeepEqual(sess.Order, []string{"a", "b"}) {
			t.Errorf("Order = %v, want [a b]", sess.Order)
		}
	})

	t.Run("does not remove already-spoken users", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a", "b", "c"})
		s.Next(ch) // current = 1 (b is speaking)
		s.Remove(ch, "a") // a already spoke — should be no-op
		sess := s.Get(ch)
		if !reflect.DeepEqual(sess.Order, []string{"a", "b", "c"}) {
			t.Errorf("Order = %v, want [a b c]", sess.Order)
		}
	})
}

func TestEnd(t *testing.T) {
	t.Run("removes the session", func(t *testing.T) {
		s := newTestStore()
		s.Start(ch, []string{"a"})
		s.End(ch)
		if s.Get(ch) != nil {
			t.Error("session should be deleted")
		}
	})

	t.Run("is a no-op for unknown channel", func(t *testing.T) {
		s := newTestStore()
		s.End("C_UNKNOWN") // should not panic
	})
}
